src/backend/distributed/commands/index.c

```diff
@@ -172,65 +188,108 @@ PlanIndexStmt(IndexStmt *createIndexStatement, const char *createIndexCommand)
+		char *namespaceName = NULL;
+#if PG_VERSION_NUM >= 120000
+		LOCKMODE lockmode = ReindexStatement->concurrent ? ShareUpdateExclusiveLock :
+							AccessExclusiveLock;
+#else
+		LOCKMODE lockmode = AccessExclusiveLock;
+#endif
+		MemoryContext relationContext = NULL;
 
-	Assert(ReindexStatement->kind == REINDEX_OBJECT_INDEX ||
-		   ReindexStatement->kind == REINDEX_OBJECT_TABLE);
+		Assert(ReindexStatement->kind == REINDEX_OBJECT_INDEX ||
+			   ReindexStatement->kind == REINDEX_OBJECT_TABLE);
 
-	/*
-	 * XXX: Consider using RangeVarGetRelidExtended with a permission
-	 * checking callback. Right now we'll acquire the lock before having
-	 * checked permissions.
-	 */
-	if (ReindexStatement->kind == REINDEX_OBJECT_INDEX)
-	{
-		Oid indOid = RangeVarGetRelid(ReindexStatement->relation,
-									  NoLock, false);
-		relation = index_open(indOid, lockmode);
-		relationId = IndexGetRelation(indOid, false);
-	}
-	else
-	{
-		relation = heap_openrv(ReindexStatement->relation, lockmode);
-		relationId = RelationGetRelid(relation);
-	}
+		if (ReindexStatement->kind == REINDEX_OBJECT_INDEX)
+		{
+			Oid indOid;
+			struct ReindexIndexCallbackState state;
+#if PG_VERSION_NUM >= 120000
+			state.concurrent = ReindexStatement->concurrent;
+#endif
+			state.locked_table_oid = InvalidOid;
+
+			indOid = RangeVarGetRelidInternal(ReindexStatement->relation,
+											  lockmode, 0,
+											  RangeVarCallbackForReindexIndex,
+											  &state);
+			relation = index_open(indOid, NoLock);
+			relationId = IndexGetRelation(indOid, false);
+		}
+		else
+		{
+			Oid tblOid = RangeVarGetRelid(ReindexStatement->relation,
+										  NoLock, false);
+			EnsureTableOwner(tblOid);
+
+			relation = heap_openrv(ReindexStatement->relation, lockmode);
+			relationId = RelationGetRelid(relation);
+		}
 
-	isDistributedRelation = IsDistributedTable(relationId);
+		isDistributedRelation = IsDistributedTable(relationId);
 
-	if (ReindexStatement->kind == REINDEX_OBJECT_INDEX)
-	{
-		index_close(relation, NoLock);
-	}
-	else
-	{
-		heap_close(relation, NoLock);
-	}
+		/*
+		 * Before we do any further processing, fix the schema name to make sure
+		 * that a (distributed) table with the same name does not appear on the
+		 * search path in front of the current schema. We do this even if the
+		 * table is not distributed, since a distributed table may appear on the
+		 * search path by the time postgres starts processing this statement.
+		 */
+		namespaceName = get_namespace_name(RelationGetNamespace(relation));
 
-	if (isDistributedRelation)
-	{
-		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg(
-							"REINDEX is not implemented for distributed relations")));
+		/* ensure we copy string into proper context */
+		relationContext = GetMemoryChunkContext(ReindexStatement->relation);
+		namespaceName = MemoryContextStrdup(relationContext, namespaceName);
+		ReindexStatement->relation->schemaname = namespaceName;
```

Comment:
I think we can skip this if schemaname != NULL. (otherwise this creates a small memory leak)

Reply:
Does that also apply to `CREATE INDEX` where I copied this code from?

Reply:
Yes. We actually added a call to copyObject before these functions are called in utility_hook.c recently, making all the memory context switching and copying redundant.

Reply:
CircleCI was segfaulting when I added the null check. I can look into this more tomorrow

Reply:
I think the segfaults were caused by my not noticing that CreateIndexStatement uses `namespaceName` later on, causing it to be left `NULL`. Also noticed we were now copying `ReindexStatement` twice, so fixed that too